core/drivers/ls_dspi.c

```diff
@@ -0,0 +1,450 @@
+	bus_setup &= ~(DSPI_CTAR_BRD | DSPI_CTAR_BRP(0x3) | DSPI_CTAR_BR(0xf));
+
+	status = dspi_convert_hz_to_baud(&req_i, &req_j, speed, bus_clock);
+
+	/* In case of failure scenario with max speed, setting default speed */
+	if (status == TEE_ERROR_ITEM_NOT_FOUND) {
+		speed = dspi_data->speed_hz;
+		EMSG("DSPI set_speed use default SCK rate %u.", speed);
+		status = dspi_convert_hz_to_baud(&req_i, &req_j,
+						 speed, bus_clock);
+	}
+
+	if (status == TEE_SUCCESS) {
+		bus_setup |= (DSPI_CTAR_BRP(req_i) | DSPI_CTAR_BR(req_j));
+		io_write32(dspi_data->base + DSPI_CTAR0, bus_setup);
+		dspi_data->speed_hz = speed;
+	} else {
+		EMSG("Unable to set speed.");
+	}
+}
+
+/* Transferred data to TX FIFO */
+static void dspi_tx(struct nxp_dspi_data *dspi_data, uint32_t ctrl,
+		    uint16_t data)
+{
+	int timeout = DSPI_TXRX_WAIT_TIMEOUT;
+	uint32_t dspi_val_addr = dspi_data->base + DSPI_PUSHR;
+	uint32_t dspi_val = ctrl | data;
+
+	/* wait for empty entries in TXFIFO or timeout */
+	while (DSPI_SR_TXCTR(io_read32(dspi_data->base + DSPI_SR)) >= 4 &&
+	       timeout--)
+		udelay(1);
+
+	if (timeout >= 0)
+		io_write32(dspi_val_addr, dspi_val);
+	else
+		EMSG("waiting timeout!");
+}
+
+/* Read data from RX FIFO */
+static uint16_t dspi_rx(struct nxp_dspi_data *dspi_data)
+{
+	int timeout = DSPI_TXRX_WAIT_TIMEOUT;
+	uint32_t dspi_val_addr = dspi_data->base + DSPI_POPR;
+
+	/* wait for valid entries in RXFIFO or timeout */
+	while (DSPI_SR_RXCTR(io_read32(dspi_data->base + DSPI_SR)) == 0 &&
+	       timeout--)
+		udelay(1);
+
+	if (timeout >= 0)
+		return (uint16_t)DSPI_RFR_RXDATA(io_read32(dspi_val_addr));
+
+	EMSG("waiting timeout!");
+	return (uint16_t)(~0);
+}
+
+static enum spi_result nxp_dspi_txrx8(struct spi_chip *chip, uint8_t *wdata,
+				      uint8_t *rdata, size_t num_pkts)
+{
+	uint8_t *spi_rd = NULL, *spi_wr = NULL;
+	uint32_t ctrl = 0;
+
+	spi_wr = wdata;
+	spi_rd = rdata;
+
+	struct nxp_dspi_data *data = container_of(chip, struct nxp_dspi_data,
+						  chip);
+	unsigned int cs = data->slave_cs;
+
+	/*
+	 * Assert PCSn signals between transfers
+	 * select which CTAR register and slave to be used for TX
+	 * CTAS selects which CTAR to be used, here we are using CTAR0
+	 * PCS (peripheral chip select) is selecting the slave.
+	 */
+	ctrl = DSPI_TFR_CTAS(data->ctar_sel) | DSPI_TFR_PCS(cs);
+	if (data->slave_mode & SPI_CONT)
+		ctrl |= DSPI_TFR_CONT;
+
+	if (data->slave_data_size_bits != 8) {
+		EMSG("data_size_bits should be 8, not %u",
+		     data->slave_data_size_bits);
+		return SPI_ERR_CFG;
+	}
+
+	while (num_pkts) {
+		if (wdata && rdata) {
+			dspi_tx(data, ctrl, *spi_wr++);
+			*spi_rd++ = dspi_rx(data);
+		} else if (wdata) {
+			dspi_tx(data, ctrl, *spi_wr++);
+			dspi_rx(data);
+		} else if (rdata) {
+			dspi_tx(data, ctrl, DSPI_IDLE_DATA);
+			*spi_rd++ = dspi_rx(data);
+		}
+		num_pkts = num_pkts - 1;
+	}
+
+	return SPI_OK;
+}
+
+static enum spi_result nxp_dspi_txrx16(struct spi_chip *chip, uint16_t *wdata,
```

Comment:
This is very similar to `nxp_dspi_txrx8()`, perhaps a helper function could be used?

Reply:
> This is very similar to `nxp_dspi_txrx8()`, perhaps a helper function could be used?

"struct spi_ops" provides support for both 8-bit and 16-bit data. So, using the same in test API to support 8-bit and 16-bit data transmission.
![image](https://user-images.githubusercontent.com/5268111/109789834-be6fac80-7c36-11eb-8c39-3b0b3be47055.png)